jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -477,6 +442,50 @@ static Chunk from(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> rele
             return new ByteBufferChunk.ReleasedByConsumer(byteBuffer, last, Objects.requireNonNull(releaser));
         }
 
+        /**
+         * <p>Creates a last/non-last Chunk with the given ByteBuffer, linked to the given Retainable.</p>
+         * <p>The {@link #retain()} and {@link #release()} methods of this Chunk will delegate to the
+         * given Retainable.</p>
+         *
+         * @param byteBuffer the ByteBuffer with the bytes of this Chunk
+         * @param last whether the Chunk is the last one
+         * @param retainable the Retainable this Chunk links to
+         * @return a new Chunk
+         */
+        static Chunk from(ByteBuffer byteBuffer, boolean last, Retainable retainable)
+        {
+            return new ByteBufferChunk.ReleasedByRetainable(byteBuffer, last, Objects.requireNonNull(retainable));
+        }
+
+        /**
+         * <p>Creates a last/non-last Chunk with the given ByteBuffer, with its own reference counter.</p>
+         * <p>This method should be used to create a Chunk that is not linked to other Retainable objects
+         * and may be passed to applications that may additionally retain/release this Chunk.</p>
+         * <p>A typical example are Chunks decoded (e.g. gzip) via {@link ContentSourceTransformer}:
+         * the decoded Chunk is not linked to the encoded Chunk, but needs its own reference counter
+         * as the decoded Chunk may in turn be {@link #slice(Chunk, int, int, boolean) sliced}.</p>
+         *
+         * @param byteBuffer the ByteBuffer with the bytes of this Chunk
+         * @param last whether the Chunk is the last one
+         * @param releaser the code to run when this Chunk is released
+         * @return a new Chunk
+         */
+        static Chunk fromWithReferenceCount(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> releaser)
```

Comment:
It is not obvious in the javadoc of the `static Chunk from(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> releaser)` method that it doesn't have a reference count.

In fact, I don't think we should ever have a chunk that does not have a real reference count (either directly or in a delegate field).

So my preference is to get rid of the other from method (which does not appear to be used) and rename this method to just `from`  (although I think `as` would be a better name for all the `from` methods here).

IF we are to have some chunks that do not have a real reference count in them, then retain must either return a boolean false OR throw an ISE.   I don't like either!

Reply:
Agreed on the reference counting argument.
I modified the `from()` methods to return either a reference counted `Chunk`, or one that wraps another `Retainable`.
Now the only use of non-reference counted `ByteBufferChunk` is for the static constants `EMPTY` and `EOF`.